README.md
# SecureStorage plugin for Apache Cordova

[![NPM](https://nodei.co/npm/cordova-plugin-secure-storage.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/cordova-plugin-secure-storage/)

## Introduction

This plugin is for use with [Apache Cordova](http://cordova.apache.org/) and allows your application to securely store secrets
such as usernames, passwords, tokens, certificates or other sensitive information (strings) on iOS & Android phones and Windows devices.

### Supported platforms

- Android
- iOS
- Windows (Windows 8.x/Store, Windows 10/UWP and Windows Phone 8.1+)

### Contents

- [Installation](#installation)
- [Plugin API](#plugin-api)
- [LICENSE](#license)

## <a name="installation"></a>Installation

Below are the methods for installing this plugin automatically using command line tools. For additional info, take a look at the [Plugman Documentation](https://cordova.apache.org/docs/en/latest/plugin_ref/plugman.html), [`cordova plugin` command](https://cordova.apache.org/docs/en/latest/reference/cordova-cli/index.html#cordova-plugin-command) and [Cordova Plugin Specification](https://cordova.apache.org/docs/en/latest/plugin_ref/spec.html).

### Cordova

The plugin can be installed via the Cordova command line interface:

- Navigate to the root folder for your phonegap project.
- Run the command:

```sh
cordova plugin add cordova-plugin-secure-storage
```

or if you want to be running the development version,

```sh
cordova plugin add https://github.com/crypho/cordova-plugin-secure-storage.git
```

## <a name="plugin_api"></a> Plugin API

#### Create a namespaced storage.

```js
var ss = new cordova.plugins.SecureStorage(
  function() {
    console.log("Success");
  },
  function(error) {
    console.log("Error " + error);
  },
  "my_app"
);
```

#### Set a key/value in the storage.

```js
ss.set(
  function(key) {
    console.log("Set " + key);
  },
  function(error) {
    console.log("Error " + error);
  },
  "mykey",
  "myvalue"
);
```

where `key` and `value` are both strings.

#### Get a key's value from the storage.

```js
ss.get(
  function(value) {
    console.log("Success, got " + value);
  },
  function(error) {
    console.log("Error " + error);
  },
  "mykey"
);
```

#### Remove a key from the storage.

```js
ss.remove(
  function(key) {
    console.log("Removed " + key);
  },
  function(error) {
    console.log("Error, " + error);
  },
  "mykey"
);
```

#### Get all keys from the storage.

```js
ss.keys(
  function(keys) {
    console.log("Got keys " + keys.join(", "));
  },
  function(error) {
    console.log("Error, " + error);
  }
);
```

#### Clear all keys from the storage.

```js
ss.clear(
  function() {
    console.log("Cleared");
  },
  function(error) {
    console.log("Error, " + error);
  }
);
```

## Platform details

#### iOS

On iOS secrets are stored directly in the KeyChain through the [SAMKeychain](https://github.com/soffes/samkeychain) library.

##### Configuration

On iOS it is possible to configure the accessibility of the keychain by setting the `KeychainAccessibility` preference in the `config.xml` to one of the following strings:

- AfterFirstUnlock
- AfterFirstUnlockThisDeviceOnly
- WhenUnlocked (default)
- WhenUnlockedThisDeviceOnly
- WhenPasscodeSetThisDeviceOnly (this option is available only on iOS8 and later)

For reference what these settings mean, see [Keychain Item Accessibility Constants](https://developer.apple.com/reference/security/keychain_services/keychain_item_accessibility_constants).

For example, include in your `config.xml` the following:

```xml
    <platform name="ios">
        <preference name="KeychainAccessibility" value="WhenUnlocked"/>
    </platform>
```

#### iOS 7 Support

iOS 7 is supported without `WhenPasscodeSetThisDeviceOnly` option.

How to test the plugin using iOS 7 simulator:

- Download and install Xcode 6 into a separate folder, e.g. /Application/Xcode 6/
- Run `$ xcode-select --switch <path to Xcode6>/Contents/Developer`
- Build Cordova app with the plugin and run it in iOS 7 simulator

#### Android

On Android there does not exist an equivalent of the iOS KeyChain. The `SecureStorage` API is implemented as follows:

- A random 256-bit AES key is generated.
- The AES key encrypts the value.
- The AES key is encrypted with a device-generated RSA (RSA/ECB/PKCS1Padding) from the Android KeyStore.
- The combination of the encrypted AES key and value are stored in `SharedPreferences`.

The inverse process is followed on `get`.

Native AES is used.
Minimum android supported version is 5.0 Lollipop. If you need to support earlier Android versions use version 2.6.8.

##### Users must have a secure screen-lock set.

The plugin will only work correctly if the user has sufficiently secure settings on the lock screen. If not, the plugin will fail to initialize and the failure callback will be called on `init()`. This is because in order to use the Android Credential Storage and create RSA keys the device needs to be somewhat secure.

In case of failure to initialize, the app developer should inform the user about the security requirements of her app and initialize again after the user has changed the screen-lock settings. To facilitate this, we provide `secureDevice` which will bring up the screen-lock settings and will call the success or failure callbacks depending on whether the user locked the screen appropriately.

For example, this would keep asking the user to enable screen lock forever. Obviously adapt to your needs :)

```js
var ss;
var _init = function() {
  ss = new cordova.plugins.SecureStorage(
    function() {
      console.log("OK");
    },
    function() {
      navigator.notification.alert(
        "Please enable the screen lock on your device. This app cannot operate securely without it.",
        function() {
          ss.secureDevice(
            function() {
              _init();
            },
            function() {
              _init();
            }
          );
        },
        "Screen lock is disabled"
      );
    },
    "my_app"
  );
};
_init();
```

##### Sharing data between 2 apps on Android.

The plugin can be used to share data securely between 2 Android apps.

This can be done by updating the `config.xml` for both the Android apps with below changes:

1. Add `xmlns:android="http://schemas.android.com/apk/res/android"` in the initial `widget` tag.
2. Add the below tag in `<platform name="android">`
```xml
<edit-config file="AndroidManifest.xml" mode="merge" target="/manifest">
    <manifest android:sharedUserId="<your>.<secret>.<anyUserId>" />
</edit-config>
```

This is required as both the apps should have the same `android:sharedUserId`. For e.g. `android:sharedUserId="com.example.myUser"`.

Consider `App1` with `packageName` as `com.test.app1`.
Data can be set from the `App1` using the below code.

```js
var ss = new cordova.plugins.SecureStorage(function() {
    ss.set(function(res) {
        console.log("Shared key set: " + res);
    }, function(err) {
        console.log("Error setting shared key: " + err);
    }, "sharedKey", "sharedValue");
}, function(err) {
    console.log("Error creating SecureStorage: " + err);
}, "my_shared_data");
```
Consider `App2` with `packageName` as `com.test.app2`.
Data can be accessed from the `App2` using the `packageName` of `App1` as shown in below code.

```js
var ss = new cordova.plugins.SecureStorage(function() {
    ss.get(function(res) {
        console.log("Got Shared key: " + res);
    }, function(err) {
        console.log("Error getting shared key: " + err);
    }, "sharedKey");
}, function(err) {
    console.log("Error creating SecureStorage: " + err);
}, "my_shared_data",
{
  android: {
      packageName: "com.test.app1"
  }
});
```

Please note that if the 2 apps use different `android:sharedUserId`, the `App2` will fail with an error `Error: Key [sharedKey] not found`.

If `App1` is uninstalled and `App2` tries to access the `sharedKey` from `App1`, `App2` will fail with an error `Error: Application package com.test.app1 not found`.

##### Android keystore deletion on security setting change

Changing the lock screen type on Android erases the keystore (issues [61989](https://code.google.com/p/android/issues/detail?id=61989) and [210402](https://code.google.com/p/android/issues/detail?id=210402)). This is also described in the [Android Security: The Forgetful Keystore](https://doridori.github.io/android-security-the-forgetful-keystore/) blog post.

This means that any values saved using the plugin could be lost if the user changes security settings. The plugin should therefore be used as a secure credential cache and not persistent storage on Android.

#### Windows

Windows implementation is based on [PasswordVault](https://msdn.microsoft.com/en-us/library/windows/apps/windows.security.credentials.passwordvault.aspx) object from the [Windows.Security.Credentials](https://msdn.microsoft.com/en-us/library/windows/apps/windows.security.credentials.aspx) namespace.
The contents of the locker are specific to the app so different apps and services don't have access to credentials associated with other apps or services.

**Limitations:** you can only store up to ten credentials per app. If you try to store more than ten credentials, you will
encounter an error. Read [documentation](https://msdn.microsoft.com/en-us/library/windows/apps/hh701231.aspx) for more details.

#### Browser

The browser platform is supported as a mock only. Key/values are stored unencrypted in localStorage.

## FAQ

- I get the error `cordova.plugins.SecureStorage is not a function`, what gives?

  You can instantiate the plugin only after the `deviceready` event is fired. The plugin is not available before that. Also make sure you use the plugin after its success callback has fired.

- Do my users really need to set a PIN code on their android devices to use the plugin?

  Yes, sorry. Android will not allow the creation of cryptographic keys unless the user has enabled at least PIN locking on the device.

## Testing

### Setup

1. Create a cordova app.
2. Change the start page in config.xml with `<content src="cdvtests/index.html" />`
3. Add your platforms.
4. Add the `cordova-plugin-test-framework` plugin:

```
cordova plugin add cordova-plugin-test-framework
```

5. Finally add the secure storage plugin as well as the tests from its location

```
cordova plugin add PATH_TO_SECURE_STORAGE_PLUGIN
cordova plugin add PATH_TO_SECURE_STORAGE_PLUGIN/tests
```

### Running the tests

Just run the app for all platforms. Remember, if you have changes to test you will need to remove the secure storage plugin and add it again for the changes to be seen by the app.

## <a name="license"></a> LICENSE

    The MIT License

    Copyright (c) 2015 Crypho AS.

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
    THE SOFTWARE.


CHANGELOG
Changelog
=========

3.0.2 - 2019-04-15
------------------

- Support data sharing between apps on android.
  [ajay-ambre]

3.0.1 - 2018-12-19
------------------

- Properly access private keys on android.
  [demetris-manikas]

3.0.0 - 2018-11-01
------------------

- Drop KitKat support, simplify android using only native AES.
  [demetris-manikas]

2.6.8 - 2017-05-29
------------------

- Fix multiple namespaced storages for android.
  [demetris-manikas]

2.6.7 - 2017-05-24
------------------

- Fixed windows bug introduced by 2.6.6.
  [digaus]

2.6.6 - 2017-05-19
------------------

- Drop unsafe iOS options.
  [demetris-manikas]

2.6.5 - 2017-04-10
------------------

- Delete saved values on new RSA key creation.
  [demetris-manikas]

2.6.4 - 2017-03-31
------------------

- Graceful handling of unsupported Android versions (4.3 and below).
  [dpa99c]

- Minor android fixes and tests.
  [demetris-manikas]

- Documentation fixes.
  [jvjvjv]

2.6.3 - 2016-11-21
------------------

- Android now thread-safe, fixes the issue with concurrent decryptions failing.
  [demetris-manikas]

2.6.2 - 2016-11-18
------------------

- Introduce a queue for decryption on Android so that only one
  happens at a time. Fixes #75.
  [ggozad]

2.6.1 - 2016-11-09
------------------

- Make sure only string values are stored.
  [demetris-manikas]

2.6.0 - 2016-10-26
------------------

- Introduce keys(), clear() methods.
  [yyfearth, demetris-manikas, ggozad]

2.5.2 - 2016-10-11
------------------

- Fix migration issues.
  [demetris-manikas]

2.5.1 - 2016-09-12
------------------

- Fix migration to SharedPreferences error in certain android phones.
  [ggozad]

2.5.0 - 2016-09-12
------------------

- Move from localStorage to SharedPreferences.
  [demetris-manikas]

2.4.2 - 2016-09-01
------------------

- Update SSKeychain for iOS to SAMKeychain.
  [briananderson1222]

- More tests.
  [ggozad]

2.4.1 - 2016-07-28
------------------

- Make intitializing SecureStorage fail on android if screen-lock settings not appropriate. Introdudce secureDevice as an android-only function to bring up screen-lock settings with only appropriate options.
  [ggozad, jdard, demetris-manikas]

2.4.0 - 2016-07-04
------------------

- Windows phone implementation.
  [sgrebnov]

- Pass javascript Error to cordova failure callbacks, cleanup.
  [sgrebnov]

2.3.1 - 2016-06-22
------------------

- Patch sjcl_ss.js to prevent cordova with browserify flag including a unnecessary huge crypto-browserify polyfill.
  [yyfearth]

2.3.0 - 2016-06-16
------------------

- Native AES on android devices that support it.
  [demetris-manikas]

2.2.4 - 2016-05-25
------------------

- Fix #30, allowing concurrent calls to securestarage in iOS,
  by not persisting callbackId's.
  [ggozad]

2.2.3 - 2016-05-18
------------------

- Fix broken version number in plugin.xml.
  [ggozad]

2.2.2 - 2016-05-10
------------------

- Execute iOS commands in separate threads instead of the main thread.
  [goshakkk / hellyeahllc]

2.2.1 - 2016-02-09
------------------

- package.json meta.
  [ggozad]

2.2.0 - 2016-01-05
------------------

- Supporting iOS 7 without 'WhenPasscodeSetThisDeviceOnly' option.
  [embedded-spirit]


2.1.0 - 2015-10-29
------------------

- Keychain accessibility setting for iOS.
  [schaumiii]


2.0.0 - 2015-10-28
------------------

- Update to use npm.
  [ggozad]

1.1.2 - 2015-04-16
------------------

- Fix Android key creation conflict.
  [demetris-manikas]

1.1.1 - 2015-04-13
------------------

- Android cleanup.
  [demetris-manikas]

1.1.0 - 2015-04-01
------------------

- Support the browser platform.
  [ggozad]

1.0.0 - 2015-03-30
------------------

- Initial release.
  [ggozad]


src/android/SecureStorage.java
package com.crypho.plugins;

import java.lang.reflect.Method;
import java.util.Hashtable;

import android.util.Log;
import android.util.Base64;
import android.os.Build;
import android.app.KeyguardManager;
import android.content.Context;
import android.content.Intent;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONException;
import org.json.JSONObject;
import org.json.JSONArray;
import javax.crypto.Cipher;

public class SecureStorage extends CordovaPlugin {
    private static final String TAG = "SecureStorage";

    private static final boolean SUPPORTED = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;

    private static final String MSG_NOT_SUPPORTED = "API 21 (Android 5.0 Lollipop) is required. This device is running API " + Build.VERSION.SDK_INT;
    private static final String MSG_DEVICE_NOT_SECURE = "Device is not secure";

    private Hashtable<String, SharedPreferencesHandler> SERVICE_STORAGE = new Hashtable<String, SharedPreferencesHandler>();
    private String INIT_SERVICE;
    private String INIT_PACKAGENAME;
    private volatile CallbackContext initContext, secureDeviceContext;
    private volatile boolean initContextRunning = false;

    @Override
    public void onResume(boolean multitasking) {
        if (secureDeviceContext != null) {
            if (isDeviceSecure()) {
                secureDeviceContext.success();
            } else {
                secureDeviceContext.error(MSG_DEVICE_NOT_SECURE);
            }
            secureDeviceContext = null;
        }

        if (initContext != null && !initContextRunning) {
            cordova.getThreadPool().execute(new Runnable() {
                public void run() {
                    initContextRunning = true;
                    try {
                        String alias = service2alias(INIT_SERVICE);
                        if (!RSA.isEntryAvailable(alias)) {
                            //Solves Issue #96. The RSA key may have been deleted by changing the lock type.
                            getStorage(INIT_SERVICE).clear();
                            RSA.createKeyPair(getContext(), alias);
                        }
                        initSuccess(initContext);
                    } catch (Exception e) {
                        Log.e(TAG, "Init failed :", e);
                        initContext.error(e.getMessage());
                    } finally {
                        initContext = null;
                        initContextRunning = false;
                    }
                }
            });
        }
    }

    @Override
    public boolean execute(String action, CordovaArgs args, final CallbackContext callbackContext) throws JSONException {
        if(!SUPPORTED){
            Log.w(TAG, MSG_NOT_SUPPORTED);
            callbackContext.error(MSG_NOT_SUPPORTED);
            return false;
        }
        if ("init".equals(action)) {
            String service = args.getString(0);
            JSONObject options = args.getJSONObject(1);
            String packageName = options.optString("packageName", getContext().getPackageName());

            Context ctx = null;
            
            // Solves #151. By default, we use our own ApplicationContext
            // If packageName is provided, we try to get the Context of another Application with that packageName
            try {
                ctx = getPackageContext(packageName);
            } catch (Exception e) {
                // This will fail if the application with given packageName is not installed
                // OR if we do not have required permissions and cause a security violation
                Log.e(TAG, "Init failed :", e);
                callbackContext.error(e.getMessage());
            }

            INIT_PACKAGENAME = ctx.getPackageName();
            String alias = service2alias(service);
            INIT_SERVICE = service;

            SharedPreferencesHandler PREFS = new SharedPreferencesHandler(alias, ctx);
            SERVICE_STORAGE.put(service, PREFS);

            if (!isDeviceSecure()) {
                Log.e(TAG, MSG_DEVICE_NOT_SECURE);
                callbackContext.error(MSG_DEVICE_NOT_SECURE);
            } else if (!RSA.isEntryAvailable(alias)) {
                initContext = callbackContext;
                unlockCredentials();
            } else {
                initSuccess(callbackContext);
            }
            return true;
        }
        if ("set".equals(action)) {
            final String service = args.getString(0);
            final String key = args.getString(1);
            final String value = args.getString(2);
            final String adata = service;
            cordova.getThreadPool().execute(new Runnable() {
                public void run() {
                    try {
                        JSONObject result = AES.encrypt(value.getBytes(), adata.getBytes());
                        byte[] aes_key = Base64.decode(result.getString("key"), Base64.DEFAULT);
                        byte[] aes_key_enc = RSA.encrypt(aes_key, service2alias(service));
                        result.put("key", Base64.encodeToString(aes_key_enc, Base64.DEFAULT));
                        getStorage(service).store(key, result.toString());
                        callbackContext.success(key);
                    } catch (Exception e) {
                        Log.e(TAG, "Encrypt failed :", e);
                        callbackContext.error(e.getMessage());
                    }
                }
            });
            return true;
        }
        if ("get".equals(action)) {
            final String service = args.getString(0);
            final String key = args.getString(1);
            String value = getStorage(service).fetch(key);
            if (value != null) {
                JSONObject json = new JSONObject(value);
                final byte[] encKey = Base64.decode(json.getString("key"), Base64.DEFAULT);
                JSONObject data = json.getJSONObject("value");
                final byte[] ct = Base64.decode(data.getString("ct"), Base64.DEFAULT);
                final byte[] iv = Base64.decode(data.getString("iv"), Base64.DEFAULT);
                final byte[] adata = Base64.decode(data.getString("adata"), Base64.DEFAULT);
                cordova.getThreadPool().execute(new Runnable() {
                    public void run() {
                        try {
                            byte[] decryptedKey = RSA.decrypt(encKey, service2alias(service));
                            String decrypted = new String(AES.decrypt(ct, decryptedKey, iv, adata));
                            callbackContext.success(decrypted);
                        } catch (Exception e) {
                            Log.e(TAG, "Decrypt failed :", e);
                            callbackContext.error(e.getMessage());
                        }
                    }
                });
            } else {
                callbackContext.error("Key [" + key + "] not found.");
            }
            return true;
        }
        if ("secureDevice".equals(action)) {
            secureDeviceContext = callbackContext;
            unlockCredentials();
            return true;
        }
        if ("remove".equals(action)) {
            String service = args.getString(0);
            String key = args.getString(1);
            getStorage(service).remove(key);
            callbackContext.success(key);
            return true;
        }
        if ("keys".equals(action)) {
            String service = args.getString(0);
            callbackContext.success(new JSONArray(getStorage(service).keys()));
            return true;
        }
        if ("clear".equals(action)) {
            String service = args.getString(0);
            getStorage(service).clear();
            callbackContext.success();
            return true;
        }
        return false;
    }

    private boolean isDeviceSecure() {
        KeyguardManager keyguardManager = (KeyguardManager)(getContext().getSystemService(Context.KEYGUARD_SERVICE));
        try {
            Method isSecure = null;
            isSecure = keyguardManager.getClass().getMethod("isDeviceSecure");
            return ((Boolean) isSecure.invoke(keyguardManager)).booleanValue();
        } catch (Exception e) {
            return keyguardManager.isKeyguardSecure();
        }
    }

    private String service2alias(String service) {
        String res = INIT_PACKAGENAME + "." + service;
        return  res;
    }

    private SharedPreferencesHandler getStorage(String service) {
        return SERVICE_STORAGE.get(service);
    }

    private void initSuccess(CallbackContext context) {
        context.success();
    }

    private void unlockCredentials() {
        cordova.getActivity().runOnUiThread(new Runnable() {
            public void run() {
                Intent intent = new Intent("com.android.credentials.UNLOCK");
                startActivity(intent);
            }
        });
    }

    private Context getContext() {
        return cordova.getActivity().getApplicationContext();
    }

    private Context getPackageContext(String packageName) throws Exception {
        Context pkgContext = null;

        Context context = getContext();
        if (context.getPackageName().equals(packageName)) {
            pkgContext = context;
        } else {
            pkgContext = context.createPackageContext(packageName, 0);
        }

        return pkgContext;
    }

    private void startActivity(Intent intent) {
        cordova.getActivity().startActivity(intent);
    }

}
